common/variant/src/main/java/org/apache/spark/types/variant/VariantShreddingWriter.java

```diff
@@ -0,0 +1,298 @@
+      int start = variantBuilder.getWritePos();
+      for (int i = 0; i < v.objectSize(); ++i) {
+        Variant.ObjectField field = v.getFieldAtIndex(i);
+        VariantSchema.ObjectField keySchema = objectSchema.get(field.key);
+        if (keySchema != null) {
+          // The field exists in the shredding schema. Recursively shred, and write the result.
+          ShreddedResult shreddedField = castShredded(field.value, keySchema.schema, builder);
+          shreddedValues[keySchema.idx] = shreddedField;
+          presentKeys.add(field.key);
+        } else {
+          // The field is not shredded. Put it in the untyped_value column.
+          int id = v.getDictionaryIdAtIndex(i);
+          fieldEntries.add(new VariantBuilder.FieldEntry(field.key, id, variantBuilder.getWritePos() - start));
+          variantBuilder.appendVariant(field.value);
+        }
+      }
+      if (presentKeys.size() != objectSchema.size()) {
+        // Set missing fields to non-null with all fields set to null.
+        // Iterate over objectSchema key-value pairs and add them to the result.
+        for (Map.Entry<String, VariantSchema.ObjectField> entry : objectSchema.entrySet()) {
+          String key = entry.getKey();
+          if (!presentKeys.contains(key)) {
+            ShreddedResult emptyChild = builder.createEmpty(entry.getValue().schema);
+            shreddedValues[entry.getValue().idx] = emptyChild;
+          }
+        }
+      }
+      result.addObject(schema, shreddedValues);
+      if (variantBuilder.getWritePos() != start) {
+        // We added something to the untyped value.
+        variantBuilder.finishWritingObject(start, fieldEntries);
+        result.addVariantValue(schema, variantBuilder.valueWithoutMetadata());
+      }
+    } else if (schema.scalarSchema != null) {
+      VariantSchema.ScalarType scalarType = schema.scalarSchema;
+      Object typedValue = tryTypedShred(v, variantType, scalarType, builder);
+      if (typedValue != null) {
+        // Store the typed value.
+        result.addScalar(schema, typedValue);
+      } else {
+        VariantBuilder variantBuilder = new VariantBuilder(false);
+        variantBuilder.appendVariant(v);
+        result.addVariantValue(schema, v.getValue());
+      }
+    } else {
+      // Store in untyped.
+      result.addVariantValue(schema, v.getValue());
+    }
+    return result;
+  }
+
+  /**
+   * Tries to cast a Variant into a typed value. If the cast fails, returns null.
+   *
+   * @param v
+   * @param variantType The Variant Type of v
+   * @param targetType The target type
+   * @return The scalar value, or null if the cast is not valid.
+   */
+  private static Object tryTypedShred(
+          Variant v,
+          VariantUtil.Type variantType,
+          VariantSchema.ScalarType targetType,
+          ShreddedResultBuilder builder) {
+    switch (variantType) {
+      case LONG:
+        if (targetType instanceof VariantSchema.IntegralType) {
```

Comment:
I think we can use the new Java grammar:
```
 if (targetType instanceof VariantSchema.IntegralType integralType) {
```

Reply:
Thanks, TIL.